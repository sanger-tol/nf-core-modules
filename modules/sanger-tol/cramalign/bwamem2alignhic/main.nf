process CRAMALIGN_BWAMEM2ALIGNHIC {
    tag "$meta.id"
    label "process_high"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/e0/e05ce34b46ad42810eb29f74e4e304c0cb592b2ca15572929ed8bbaee58faf01/data' :
        'community.wave.seqera.io/library/bwa-mem2_htslib_samtools:db98f81f55b64113' }"

    input:
    tuple val(meta), path(cram), path(crai), val(chunkn), val(range), path(index), path(reference)

    output:
    tuple val(meta), path("*.bam"), emit: bam
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args1 = task.ext.args1 ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def args4 = task.ext.args4 ?: ''
    def args5 = task.ext.args5 ?: ''
    def args6 = task.ext.args6 ?: ''
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    // Please be aware one of the tools here required mem = 28 * reference size!!!
    """
    INDEX=`find -L ./ -name "*.amb" | sed 's/\\.amb\$//'`

    samtools cat ${args1} -r "#:${range[0]}-${range[1]}" ${cram} |\\
        samtools fastq ${args2} - |\\
        bwa-mem2 mem ${args3} -t ${task.cpus} \${INDEX} - |\\
        samtools fixmate ${args4} - - |\\
        samtools view -h ${args5} |\\
        samtools sort ${args6} -@${task.cpus} -T ${prefix}_tmp -o ${prefix}.bam -

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        bwamem2: \$(echo \$(bwa-mem2 version 2>&1) | sed 's/.* //')
    END_VERSIONS
    """

    stub:
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        bwamem2: \$(echo \$(bwa-mem2 version 2>&1) | sed 's/.* //')
    END_VERSIONS
    """
}
