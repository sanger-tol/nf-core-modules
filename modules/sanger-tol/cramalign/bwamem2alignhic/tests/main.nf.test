
nextflow_process {

    name "Test Process CRAMALIGN_BWAMEM2ALIGNHIC"
    script "../main.nf"
    process "CRAMALIGN_BWAMEM2ALIGNHIC"

    tag "modules"
    tag "modules_sangertol"
    tag "bwamem2/index"
    tag "cramalign"
    tag "cramalign/bwamem2alignhic"
    tag "nf-core/bwamem2/index"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall("${launchDir}/library/", ["bwamem2/index"])
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")

        run("BWAMEM2_INDEX") {
            script "../../../../nf-core/bwamem2/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id: "test" ],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("meles meles") {
        config "./nextflow.config"

        when {

            params {
                samtools_cat_args = ""
                samtools_fastq_args = "-F0xB00 -nt"
                bwamem2_mem_args = "-5SPp"
                samtools_fixmate_args = "-mpu"
                samtools_view_args = "-q 0 -F 0x904"
                samtools_sort_args = "--write-index -l1"
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true),
                ]).join(BWAMEM2_INDEX.out.index)
                .map { meta, cram, crai, chunkn, range, ref, index ->
                    [ meta, cram, crai, chunkn, range, index, ref ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        bam(process.out.bam.get(0).get(1)).getReadsMD5(),
                        process.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/")
        }

    }

    test("meles meles - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true),
                ]).join(BWAMEM2_INDEX.out.index)
                .map { meta, cram, crai, chunkn, range, ref, index ->
                    [ meta, cram, crai, chunkn, range, index, ref ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/")
        }

    }

}
