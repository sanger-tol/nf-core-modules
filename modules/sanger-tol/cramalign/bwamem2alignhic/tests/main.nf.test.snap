{
    "meles meles": {
        "content": [
            "12bbc3437515580a7dcc3c0d677c225a",
            [
                "versions.yml:md5,94aadbb19ae8ccacf115ea785735a3b0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-03T10:00:52.296785"
    },
    "meles meles - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,94aadbb19ae8ccacf115ea785735a3b0"
                ],
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,94aadbb19ae8ccacf115ea785735a3b0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-03T10:01:16.885815"
    }
}